### PR TITLE
Add FATES cross-grid seed dispersal mechanism

### PR DESCRIPTION
**Pull request by glemieux**
This PR enables FATES to disperse seeds across neighboring grid cells using MPI.  The API update includes calls to new fates dispersal procedures and as such should be coordinated with https://github.com/NGEET/fates/pull/1005.

[B4B] except for FATES

### CODE REVIEW

**Comment by bishtgautam** [comment]
@glemieux Did you forget to update the FATES submodule?

**Comment by bishtgautam** [comment]
Btw, this is a very exciting PR and I'm interested to find out the approach you and the FATES team took for this work.

File: components/elm/src/main/elmfates_interfaceMod.F90 [hunk 1]
```diff
@@ -2400,6 +2444,166 @@ subroutine wrap_canopy_radiation(this, bounds_clump, &
...
+    if (fates_dispersal_kernel_mode .eq. fates_dispersal_kernel_none) return 
+   
+    nc = bounds_clump%clump_index
+
+    do s = 1, this%fates(nc)%nsites
+       c = this%f2hmap(nc)%fcolumn(s)
+       g = col_pp%gridcell(c)
+       ! loop over pft. Disperse seeds for pft = 9 for now
+       !write(iulog,*) 's, c, g: ', s, c, g
+       !write(iulog,*) 'BEFORE, this%fates(nc)%bc_in(s)%seed_in(9),this%fates(nc)%bc_out(s)%seed_out(9): ', this%fates(nc)%bc_in(s)%seed_in(9), this%fates(nc)%bc_out(s)%seed_out(9)
```
Comment (bishtgautam) [inline components/elm/src/main/elmfates_interfaceMod.F90:2595]: please delete the commented out code.

File: components/elm/src/main/elmfates_interfaceMod.F90 [hunk 2]
```diff
@@ -2400,6 +2444,166 @@ subroutine wrap_canopy_radiation(this, bounds_clump, &
...
+
+       this%fates(nc)%bc_in(s)%seed_in(:) = 0 ! reset 
+
+       !write(iulog,*) 'AFTER, this%fates(nc)%bc_in(s)%seed_in(9),this%fates(nc)%bc_out(s)%seed_out(9): ', this%fates(nc)%bc_in(s)%seed_in(9), this%fates(nc)%bc_out(s)%seed_out(9)
```
Comment (bishtgautam) [inline components/elm/src/main/elmfates_interfaceMod.F90:2599]: please delete the commented out code.

File: components/elm/src/main/elmfates_interfaceMod.F90 [hunk 3]
```diff
@@ -2400,6 +2444,166 @@ subroutine wrap_canopy_radiation(this, bounds_clump, &
...
+
+    nc = bounds_clump%clump_index
+    
+    ! Add fates check for seed dispersal mode
+    do s = 1, this%fates(nc)%nsites
+       c = this%f2hmap(nc)%fcolumn(s)
+       g = col_pp%gridcell(c)
+
+       ! Check that it is the beginning of the current dispersal time step
+       !write(iulog, *) 'wrapseeddisp: isittime, hlm_day: ', IsItDispersalTime(), hlm_current_day
```
Comment (bishtgautam) [inline components/elm/src/main/elmfates_interfaceMod.F90:2545]: please delete the commented out code.

File: components/elm/src/main/elmfates_interfaceMod.F90 [hunk 4]
```diff
@@ -2400,6 +2444,166 @@ subroutine wrap_canopy_radiation(this, bounds_clump, &
...
+       if (IsItDispersalTime()) then
+       
+          ! Future work: Determine the number of fates sites in the gridcell and distribute over fates sites on gridcell
+          !write(iulog,*) 's, c, g, incoming: ', s, c, g, sum(this%fates_seed%incoming_global(g,:))
```
Comment (bishtgautam) [inline components/elm/src/main/elmfates_interfaceMod.F90:2549]: please delete the commented out code.

File: components/elm/src/main/elmfates_interfaceMod.F90 [hunk 5]
```diff
@@ -2400,6 +2444,166 @@ subroutine wrap_canopy_radiation(this, bounds_clump, &
...
+
+          ! need to devide seed_id_global by the number of sites in one grid
+          this%fates(nc)%bc_in(s)%seed_in(:) = this%fates_seed%incoming_global(g,:)   !/this%fates(nc)%nsites ! assuming equal area for all sites, seed_id_global in [kg/grid/day], seed_in in [kg/site/day]
+          this%fates(nc)%bc_out(s)%seed_out(:) = 0._r8             ! reset seed_out
+
+          !write(iulog,*) 's, c, g, bc_in: ', s, c, g, sum(this%fates(nc)%bc_in(s)%seed_in(:))
```
Comment (bishtgautam) [inline components/elm/src/main/elmfates_interfaceMod.F90:2555]: please delete the commented out code.

File: components/elm/src/main/elmfates_interfaceMod.F90 [hunk 6]
```diff
@@ -2400,6 +2444,166 @@ subroutine wrap_canopy_radiation(this, bounds_clump, &
...
+                                                 this%fates_seed%outgoing_global(neighbor%gindex,:) * &
+                                                 neighbor%density_prob(:) / lneighbors(g)%neighbor_count
+
+            !write(iulog,*) 'g, gi, densprob: ', g, neighbor%gindex, sum(neighbor%density_prob(:))
+            !write(iulog,*) 'g, gi, ncount: ', g, neighbor%gindex, lneighbors(g)%neighbor_count
+
+            neighbor => neighbor%next_neighbor
+         end do
+         !write(iulog,*) 'g, outgoing: ', g, sum(this%fates_seed%outgoing_global(g,:))
+         !write(iulog,*) 'g, incoming: ', g, sum(this%fates_seed%incoming_global(g,:))
```
Comment (bishtgautam) [inline components/elm/src/main/elmfates_interfaceMod.F90:2508]: please delete the commented out code.

File: components/elm/src/main/elmfates_interfaceMod.F90 [hunk 7]
```diff
@@ -2400,6 +2444,166 @@ subroutine wrap_canopy_radiation(this, bounds_clump, &
...
+   integer :: g    ! gridcell index
+       
+   type (neighbor_type), pointer :: neighbor
+   
+   ! Check if seed dispersal mode is 'turned on', if not return to calling procedure
+   if (fates_dispersal_kernel_mode .eq. fates_dispersal_kernel_none) return 
+   
+   call t_startf('fates-seed-mpi_reduce')
+
+   !write(iulog, *) 'wrapglobal: isittime, hlm_day: ', IsItDispersalTime(), hlm_current_day
```
Comment (bishtgautam) [inline components/elm/src/main/elmfates_interfaceMod.F90:2472]: please delete the commented out code.

File: components/elm/src/main/elmfates_interfaceMod.F90 [hunk 8]
```diff
@@ -1183,9 +1213,23 @@ subroutine wrap_update_hlmfates_dyn(this, nc, bounds_clump,      &
...
+          ! zero the outgoing seed array
+          this%fates_seed%outgoing_local(:,:) = 0._r8
+       end if
+       
        do s = 1,this%fates(nc)%nsites
 
           c = this%f2hmap(nc)%fcolumn(s)
+          g = col_pp%gridcell(c)
+          
+          !write(iulog, *) 'wrapupdate: isittime, hlm_day: ', IsItDispersalTime(), hlm_current_day
```
Comment (bishtgautam) [inline components/elm/src/main/elmfates_interfaceMod.F90:1227]: please delete the commented out code.

**Comment by glemieux** [comment]
> @glemieux Did you forget to update the FATES submodule?

I was going to push the final update once we've integrated https://github.com/NGEET/fates/pull/1005 so I can point to the specific fates `main` commit.  For now I was planning on only updating locally when I run the land developer tests.  I'm more than happy to push that interim update though.

File: components/elm/src/main/elmfates_interfaceMod.F90 (hunk 1 above)
Comment (glemieux) [reply components/elm/src/main/elmfates_interfaceMod.F90]: Thanks for catching these!

File: components/elm/src/main/elmfates_interfaceMod.F90 (hunk 2 above)
Comment (glemieux) [reply components/elm/src/main/elmfates_interfaceMod.F90]: done

**Comment by bishtgautam** [comment]
I'll be happy to upload the new parameter file on the inputdata server.

**Comment by rljacob** [comment]
status:  waiting on a corresponding FATES PR.

**Comment by rljacob** [comment]
status:  still waiting on a FATES PR.

**Comment by glemieux** [comment]
@bishtgautam this will need to come in after the API25.5 update in https://github.com/E3SM-Project/E3SM/pull/5802.

**Comment by rljacob** [comment]
notes:  should be done soon.  FATES-side tests are progressing.

**Comment by rljacob** [comment]
Waiting on #5849

File: components/elm/src/main/elmfates_interfaceMod.F90 (hunk 3 above)
Comment (glemieux) [reply components/elm/src/main/elmfates_interfaceMod.F90]: complete

File: components/elm/src/main/elmfates_interfaceMod.F90 (hunk 4 above)
Comment (glemieux) [reply components/elm/src/main/elmfates_interfaceMod.F90]: complete

File: components/elm/src/main/elmfates_interfaceMod.F90 (hunk 6 above)
Comment (glemieux) [reply components/elm/src/main/elmfates_interfaceMod.F90]: done

File: components/elm/src/main/elmfates_interfaceMod.F90 (hunk 8 above)
Comment (glemieux) [reply components/elm/src/main/elmfates_interfaceMod.F90]: complete

File: components/elm/src/main/elmfates_interfaceMod.F90 (hunk 5 above)
Comment (glemieux) [reply components/elm/src/main/elmfates_interfaceMod.F90]: complete

File: components/elm/src/main/elmfates_interfaceMod.F90 (hunk 7 above)
Comment (glemieux) [reply components/elm/src/main/elmfates_interfaceMod.F90]: complete

**Comment by glemieux** [comment]
@peterdschwartz now that #5849 is merged, we can start moving forward with this again.  I'm working on an ERS fix on the fates-side that accompanies this, but that should be done in the next day or so.  After that I'll rebase and retest this.

**Comment by rljacob** [comment]
status:  working on an exact restart issue.

**Comment by rljacob** [comment]
@glemieux any update on restart issue?

**Comment by glemieux** [comment]
> @glemieux any update on restart issue?

@rljacob There is an issue with the PEM and ERP for this.  The current plan is to make this a known issue and integrate into fates main/hlm master branches and address this later.

**Comment by rljacob** [comment]
update: currently being worked on.

**Comment by glemieux** [comment]
Regression testing against master on `pm-cpu` is in progress.  The `e3sm_land_debug` tests are complete and coming back as b4b against `master` with two exceptions:
- `SMS_Ld20_D.f45_f45.IELMFATES.pm-cpu_gnu.elm-fates_rd`
- `ERS_D.ne11_oQU240.I20TRELM.pm-cpu_gnu`

The differences in the fates testmod are expected as this PR includes a number of known answer changing scientific updates for fates science.  The non-fates test is only showing DIFFs in the `cpl`.  @peterdschwartz is this expected?

Location on perlmutter: `/pscratch/sd/g/glemieux/e3sm-tests/pr5555-eld-debug.fates.pm-cpu.gnu.C9a524415e4-F1d18bb69.DEBUG`

**Comment by glemieux** [comment]
The `e3sm_land_developer` regression tests are complete and showing similar results to the debug tests.  Nearly everything is b4b except for the fates testmods, which is as expeced, and the `ERS_D.ne11_oQU240.I20TRELM.pm-cpu_gnu` testmod is again coming back with a DIFF against the `master` baseline.

There is also a fates test that is not passing exart restart, but this is expected and will be corrected in a forthcoming PR.

Location on perlmutter: `/pscratch/sd/g/glemieux/e3sm-tests/pr5555-eld-nodebug.fates.pm-cpu.gnu.C9a524415e4-F1d18bb69`

**Comment by glemieux** [comment]
Testing with https://github.com/NGEET/fates/pull/1109/files in progress to address known ERS issue.

**Comment by glemieux** [comment]
@peterdschwartz FATES restart issue has been resolved via https://github.com/E3SM-Project/E3SM/pull/5555/commits/e7921a0c0d7b7234dc6d06f52ae59785116f8715.  I rebased against `master` and I'm rerunning `e3sm_land_developer` regression tests on perlmutter.

**Comment by glemieux** [comment]
@peterdschwartz this is finally all good to go.  Regression tests on perlmutter are coming back B4B with the known exception of the `fates` testmods.  

Regression test results: `/global/homes/g/glemieux/scratch/e3sm-tests/pr5555-pr1114-eld-intel.fates.pm-cpu..Ce7921a0c0d-F1874511a`

**Comment by peterdschwartz** [comment]
merged to next

**Review by peterdschwartz — APPROVED** [review]

**Pull request merged**